### ISSUE TITLE
Repeated samples in simple DAS query

### ISSUE THREAD

**Issue by jfernan2** [vkuznet/dasgoclient#2]
Thanks for this extra fast tool!!
However, a simple dataset site=T3_* query seems to give repeated results:

 /afs/cern.ch/user/v/valya/public/dasgoclient/dasgoclient_linux -query="dataset site=T3_ES_Oviedo"

/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TT_13TeV-amcatnlo-herwigpp/RunIIWinter15wmLHE-MCRUN2_71_V1-v4/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE
/TTTT_TuneCUETP8M1_13TeV-amcatnlo-pythia8/RunIIWinter15wmLHE-MCRUN2_71_V1_ext1-v1/LHE

Thanks!

**Comment by vkuznet** [comment]
Hi,
thanks for report, so far I concentrated on speed rather a convenience.
Of course sorting can be implemented, but when results will grow, e.g.
dataset at T1 site, the speed will decrease if sorting is done.

You can easily achieve it anyway using standard UNIX tools:
dasgoclient -query="dataset site=T3_ES_Oviedo" | sort | uniq

I'll keep ticket open and think if it worth to implement the sorting.
Best,
Valentin.

**Comment by jordantucker** [comment]
Hi Valentin (from down the hall),
I also like the tool--so much faster!
For the query
site dataset=/MinBias_TuneCUETP8M1_13TeV-pythia8/RunIISummer15GS-MCRUN2_71_V1-v2/GEN-SIM
I got 93473 lines, of which only 54 were unique--a bit annoying over remote connection to have that many lines scroll my screen.
sort|uniq is fine, and I can even wrap in a shell function, but why have to? :)
Thanks!
Jordan

**Comment by vkuznet** [comment]
Jordan,
this query maps into Phedex call which find all replicas of the given dataset,
that's why it print that many lines. Now, the response I get is always a full
one, e.g. json, out of which I strip site names only in default mode.

As I mention before since I don't know want to have a knowledge of the query
I should carry about generic output. As such a query 
file dataset=/a/b/c
can easily yield O(100K) UNIQUE files. And, I didn't decided yet
if it's worth to apply sort for all queries (generic use case).

For site queries it makes sense, but for other queries, e.g. files, does not.
Therefore introducing sort in a tool will be useless for majority of the
queries (basically for all except site one), which as you can imaging
will introduce latency overhead.

But I'll check how much time sort takes for O(100K), O(1M) strings
and decide if it's worth to implement it.

Best,
Valentin.

**Comment by vkuznet** [comment]
Done

**Comment by vkuznet** [comment]
Jordan,
I fixed this issue and now dasgoclient reports unique entries.
Feel free to try it out from the same AFS location.
Best,
Valentin.

**Comment by jordantucker** [comment]
Works great, thanks very much! -J

**Comment by vkuznet** [comment]
Hi, I realized that internal sorting/uniqueness significantly slows down performance on very large queries, e.g.
```
file dataset=/Neutrino_E-10_gun/RunIISpring15PrePremix-PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v2-v2/GEN-SIM-DIGI-RAW
```
returns 125K+ files. If sorting and check for result's uniqueness is applied the query runs 20x slower (2min vs 6sec when no sorting is applied). Therefore I decided to make this behavior optional via new `-unique` option. Those users who care about unique results will need to use this option explicitly, by default it is set to false to allow fastest response time. The changes will go in v00.00.06 tag and eventually will be propagated into cvmfs where dasgoclient will reside.
